#line136	nonsynonymous SNV	SAMD11:NM_152486:exon7:c.T690G:p.H230Q,	1	874824	874824	T	G	hom	1	874824	.	TG	2	mf1	DP=25;MQ0=0;SOMATIC;SS=Somatic;VT=SNP;VLSC=.;ANNOVAR_DATE=2017-06-01;Func.refGene=exonic;Gene.refGene=SAMD11;
#GeneDetail.refGene=.;ExonicFunc.refGene=nonsynonymous_SNV;AAChange.refGene=SAMD11:NM_152486:exon7:c.T690G:p.H230Q;Func.knownGene=exonic;Gene.
#knownGene=SAMD11;GeneDetail.knownGene=.;ExonicFunc.knownGene=nonsynonymous_SNV;AAChange.knownGene=SAMD11:uc031pko.1:exon1:c.T135G:p.H45Q,SAMD
#11:uc031pke.1:exon5:c.T135G:p.H45Q,SAMD11:uc031pjm.1:exon6:c.T693G:p.H231Q,SAMD11:uc031pju.1:exon6:c.T693G:p.H231Q,SAMD11:uc001abw.1:exon7:c.
#T690G:p.H230Q,SAMD11:uc031pjv.1:exon7:c.T135G:p.H45Q;cytoBand=1p36.33;1000g2015aug_all=.;ExAC_ALL=.;ExAC_AFR=.;ExAC_AMR=.;ExAC_EAS=.;ExAC_FIN
#=.;ExAC_NFE=.;ExAC_OTH=.;ExAC_SAS=.;avsnp147=.;SIFT_score=0.183;SIFT_pred=T;Polyphen2_HDIV_score=0.008;Polyphen2_HDIV_pred=B;Polyphen2_HVAR_s
#core=0.001;Polyphen2_HVAR_pred=B;LRT_score=.;LRT_pred=.;MutationTaster_score=1;MutationTaster_pred=N;MutationAssessor_score=0.69;MutationAsse
#ssor_pred=N;FATHMM_score=.;FATHMM_pred=.;PROVEAN_score=-0.96;PROVEAN_pred=N;VEST3_score=0.309;CADD_raw=-0.207;CADD_phred=1.064;DANN_score=0.6
#32;fathmm-MKL_coding_score=0.058;fathmm-MKL_coding_pred=N;MetaSVM_score=-1.009;MetaSVM_pred=T;MetaLR_score=0.058;MetaLR_pred=T;integrated_fit
#Cons_score=0.598;integrated_confidence_value=0;GERP++_RS=0.619;phyloP7way_vertebrate=0.062;phyloP20way_mammalian=-0.202;phastCons7way_vertebr
#ate=0.005;phastCons20way_mammalian=0.002;SiPhy_29way_logOdds=.;ALLELE_END	GT:AD:DP:FA:MQ0:BQ:SS:SSC	0/0:16,0:16:0.000:0:.:2:.	0/1:8,1:9:0.111:0:31:2:.




#There is no AD, RD, FA
#There is BCOUNT

#line115	stopgain	MEGF6:NM_001409:exon6:c.C618A:p.C206X,	chr1	3432078	3432078	C	T	hom	chr1	3432078	.	C	T.	PASS	CSQ=T|intron_variant|MODIFIER|PRDM16|ENSG00000142611|Transcript|ENST00000514189|protein_coding||15/15|ENST00000514189
#.4:c.3521+970C>T||-/4236|-/3534|-/1177||||1||1||SNV|HGNC|HGNC:14000||5|||ENSP00000421400||D6RFY3|UPI0001D3BC63|||||||||||||||||||||||||||||||
#|63976|,T|missense_variant|MODERATE|PRDM16|ENSG00000142611|Transcript|ENST00000270722|protein_coding|16/17||ENST00000270722.8:c.3634C>T|ENSP0
#0000270722.5:p.Leu1212Phe|3683/8690|3634/3831|1212/1276|L/F|Ctt/Ttt||1||1||SNV|HGNC|HGNC:14000|YES|1||CCDS41236.2|ENSP00000270722|Q9HAZ2||UPI
#0000458A29|NM_022114.3||tolerated(1)|benign(0.003)||||||||||||||||||||||||||||63976|,T|missense_variant|MODERATE|PRDM16|ENSG00000142611|Trans
#cript|ENST00000378391|protein_coding|16/17||ENST00000378391.5:c.3634C>T|ENSP00000367643.2:p.Leu1212Phe|3697/5447|3634/3774|1212/1257|L/F|Ctt/
#Ttt||1||1||SNV|HGNC|HGNC:14000||1||CCDS44048.2|ENSP00000367643|Q9HAZ2||UPI0000458A2A|NM_199454.2||tolerated_low_confidence(1)|benign(0.042)||
#||||||||||||||||||||||||||63976|,T|intron_variant|MODIFIER|PRDM16|ENSG00000142611|Transcript|ENST00000511072|protein_coding||15/15|ENST000005
#11072.4:c.3524+970C>T||-/4282|-/3537|-/1178||||1||1||SNV|HGNC|HGNC:14000||5|||ENSP00000426975||D6RDW0|UPI0001D3BC62||||||||||||||||||||||||||
#||||||63976|,T|intron_variant&non_coding_transcript_variant|MODIFIER|PRDM16|ENSG00000142611|Transcript|ENST00000378389|processed_transcript||
#4/4|ENST00000378389.5:n.723+970C>T||-/1142||||||1||1||SNV|HGNC|HGNC:14000||5||||||||||||||||||||||||||||||||||||||63976|,T|missense_variant|M
#ODERATE|PRDM16|ENSG00000142611|Transcript|ENST00000509860|protein_coding|12/13||ENST00000509860.1:c.3058C>T|ENSP00000425796.1:p.Leu1020Phe|30
#58/3737|3058/3255|1020/1084|L/F|Ctt/Ttt||1||1|cds_start_NF|SNV|HGNC|HGNC:14000||5|||ENSP00000425796||H0YA13|UPI0001D3BC64|||tolerated(1)|beni
#gn(0.005)||||||||||||||||||||||||||||63976|,T|non_coding_transcript_exon_variant&non_coding_transcript_variant|MODIFIER|PRDM16|ENSG0000014261
#1|Transcript|ENST00000512462|processed_transcript|15/16||ENST00000512462.4:n.3412C>T||3412/3723||||||1||1||SNV|HGNC|HGNC:14000||1||||||||||||
#||||||||||||||||||||||||||63976|	GT:IGT:DP:DP4:BCOUNT:GQ:JGQ:VAQ:BQ:MQ:AMQ:SS:SSC	0/0:0/0:186:64,119,1,2:2,183,0,1:255:163:0:28
#:60:60:0:.	0/1:0/1:87:20,40,10,17:0,60,1,26:156:163:156:27,30:60:60,60:2:163



#!/usr/bin/python
import sys, os, subprocess, string

gzip = 'zcat'


def global_vars():
	# Annovar ouput positions
	global nchr, nid, npos, nref, nalt, npass, ndat, nger, ntum, ninfo
	nchr = 0
	nid = 2
	npos = 1
	nref = 3
	nalt = 4
	npass = 6
	ninfo = 7
	ndat = 8
	nger = 9
	ntum = 10
	global ptype, info_ann, info_start, info_stop
	ptype = 1
	info_ann = 2
	info_start = 9
	info_stop = 18
	global ann_chr, ann_id, ann_pos, ann_ref, ann_alt, ann_pass, ann_tdat, ann_ger, ann_tum, ann_info
	ann_chr = info_start + nchr
	ann_id = info_start + nid
	ann_pos = info_start + npos
	ann_ref = info_start + nref
	ann_alt = info_start + nalt
	ann_pass = info_start + npass
	# Important position in the vcf with information
	ann_tdat = info_start + ndat
	ann_ger = info_start + nger
	ann_tum = info_start + ntum
	ann_info = info_start + ninfo
	return



def get_mapping_data(data_types, data_values):
	dict = {}
	vtype = data_types.split(':')
	if 'GT' not in vtype:
		print 'ERROR: Incorrect FORMAT position', ann_tdat + 1, 'in VCF'
		exit()
	vdata = data_values.split(':')
	if (1 or len(vtype) == len(vdata)):
		for i in range(len(vtype)):
			if vtype[i] == 'GT':
				if vdata[i].find('/'):
					dict[vtype[i]] = vdata[i].split('/')
				elif vdata[i].find('|'):
					dict[vtype[i]] = vdata[i].split('|')
				else:
					dict[vtype[i]] = [vdata[i]]
			elif vtype[i] == 'AD':
				vnum = []
				v = vdata[i].split(',')
				for num in v:
					try:
						vnum.append(float(num))
					except:
						vnum.append(0.0)
				dict[vtype[i]] = vnum
			elif vtype[i] == 'BQ':
                                dict[vtype[i]] = vdata[i].split(',')[0]
				
			else:
				try:
					dict[vtype[i]] = float(vdata[i])
				except:
					dict[vtype[i]] = 0.0
	return dict

def get_frequency(data_info):
        dict = {}
        info = data_info.split(';')
        #print info
        for i in range(len(info)):
                infoFreq = info[i].split('=')
                #print infoFreq
                if infoFreq[0] ==  'gnomAD_genome_ALL':
                        if infoFreq[1] == '.': infoFreq[1] = '1.1' #Avoid '.' (not find)
                        else: dict[infoFreq[0]]=float(infoFreq[1])
                elif infoFreq[0] == 'ExAC_ALL':
                        if infoFreq[1] == '.': infoFreq[1] = '1.1' #Avoid '.' (not find)
                        else: dict[infoFreq[0]]=float(infoFreq[1])
                elif infoFreq[0] == '1000g2015aug_all':
                        if infoFreq[1] == '.': infoFreq[1] = '1.1' #Avoid '.' (not find)
                        else: dict[infoFreq[0]]=float(infoFreq[1])
                #else:
                        #print "ERROR: AF field not found"

        return dict

def get_gene(data_info): # for COAD
        info = data_info.split(':')
        return info[0]


def get_annovar(path, filelist, pos, rpos=-1, tfdp=0, tfaval=0.0, tbq=0, fpass=True, both=False, normFilt=False, maf=1.0, filtMut=False, minAD=0, minRatioAD_DP=0):
	vmuts = []
	# rpos, position of reference = tumor values. Filtered out if bad quality for tumor
        #pos, position of normal values
        #print "pos=",pos,"rpos=",rpos,"both=",both
        #print "fdp=",tfdp,"faval=",tfaval,"tbq=",tbq, "fpass=",fpass

	with open(filelist, 'r') as list_file:
         for lineList in list_file:
           NameList=lineList.split()
	   filename=path+'/'+NameList[0] +'.exonic_variant_function'   # Now the file list has three columns. Only the first one
#	   filename=path+'/'+lineList.rstrip()+'.exonic_variant_function'
	   fileout=filename+'_clean'
	   with open(filename, 'r') as f:
		l = []
		for line in f:
			v = line.split('\t')
			if len(v) > pos:
				chr = v[ann_chr]
				#rs = v[ann_id]
				rs = '.'
				ipos = v[ann_pos]
                                aref=v[ann_ref]
                                aalt=v[ann_alt]
                                #snp=(chr,rs,ipos,aref,aalt)
				aa = v[ann_ref] + '|' + v[ann_alt]
				first_data = v[info_ann].split(',')[0]
				data = first_data.split(':')
				#print data,len(data)
				gene = data[0]   # for COAD
				#gene = get_gene(v[ann_info])
				if (len(data) >= 5): mut = data[4]
			filterPass = v[ann_pass]
			idata = v[pos]
			
			#Filter for chrm X
			#if chr == 'X' or chr == 'chrX': continue 
			#print "Filtered by chrm ", chr
			
			#Get frequency data from Annovar file: gnomAD_genome_ALL, ExAC_ALL, 1000g2015aug_all
                        dic_info_freq = get_frequency(v[ann_info])
                        #print dic_info_freq
                        if not bool(dic_info_freq):
                                #print 'Frequency dictionary is empty'
                                freqMin = 0
                                freqMax = 0.0
                        else:
                                freqMin = min(dic_info_freq.get('gnomAD_genome_ALL', 1.1), dic_info_freq.get('ExAC_ALL', 1.1), dic_info_freq.get('1000g2015aug_all', 1.1))
                        	freqMax = max(dic_info_freq.get('gnomAD_genome_ALL', 0), dic_info_freq.get('ExAC_ALL', 0), dic_info_freq.get('1000g2015aug_all', 0))
			#print "Min freq is: ", freq
                        freqMaxstr=str(freqMax)


			#print "v[ann_tdat]=",v[ann_tdat], " ",ann_tdat,"\n"
			#print "v[pos]=",v[pos], " ",pos,"\n"
			dic_data = get_mapping_data(v[ann_tdat], v[pos])
			#print "v[rpos]=",v[rpos], " ", rpos, "\n" 
			dic_data_ref = get_mapping_data(v[ann_tdat], v[rpos])
			if both:
					bqval = min(dic_data.get('BQ', 0), dic_data_ref.get('BQ', 0))
					faval = min(dic_data.get('FA', 0), dic_data_ref.get('FA', 0))
					fdp = min(dic_data.get('DP', 0), dic_data_ref.get('DP', 0))
			elif normFilt:
					bqval = dic_data.get('BQ', 0)
                                        faval = dic_data.get('FA', 0)
                                        fdp = dic_data.get('DP', 0)	
			else:
					bqval = dic_data_ref.get('BQ', 0)
					faval = dic_data_ref.get('FA', 0)
					fdp = dic_data_ref.get('DP', 0)
			
			#print dic_data.get('BQ', 0)
			#print dic_data_ref.get('BQ', 0)
			#print "filterPass=",filterPass, "fdp",fdp, "faval=",faval, "bqval=",bqval
                        #if bqval < tbq :  continue
                        #if faval < tfaval : continue 
                        #if fdp < tfdp : continue
			fpass = False
			if fpass == True and filterPass != 'PASS': continue
			#print "Ha pasado los filtros bq, fdp y PASS"

			#Filter for AF
                        if (float(freqMax) > maf):
                                #print 'Variant ' + str(pos) + ' filtered by freq: ' + str(freqMax) + ' in filename: ' + filename
                                continue

			#Get the Genotype:
			iaa = dic_data['GT']
			iaar = dic_data_ref['GT']
                        
			#Get the DP and AD for normal and tumor
			fdp = dic_data.get('DP', 0)
                        fdpr = dic_data_ref.get('DP', 0)

                        ad_alt = dic_data.get('AD',[0,0])[1]
                        ad_ref = dic_data.get('RD',0)
			#print 'RD= ' + str(ad_ref) + ' and AD= ' + str(ad_alt)
			if (fdp>0) fdpnor=fdp
			else: fdpnor = ad_alt + ad_ref
			if (fdpnor==0):
                                #print 'AD,RD '+str(ad_alt) + ','+str(ad_ref)+':'+filename
                                #snp=(chr,rs,ipos,aref,aalt,freqMaxstr)
                                #print snp
                                ratio_ad=0
                        else :
                                ratio_ad = ad_alt / fdpnor

                        adr_alt = dic_data_ref.get('AD',[0,0])[1]
                        adr_ref = dic_data_ref.get('RD',0)
			if (fdpr>0) fpdtum=fdpr
			else: fdptum = adr_alt + adr_ref
			if (fdptum==0):
                                #print 'ADt,RDt '+str(adr_alt) + ','+str(adr_ref)+':'+filename
                                #snp=(chr,rs,ipos,aref,aalt,freqMaxstr)
                                #print snp
                                ratio_adr=0
                        else :
                                ratio_adr = adr_alt / fdptum

			#print 'minAD= ' + str(minAD) + ' and minRatioAD_DP= ' + str(minRatioAD_DP)
			if ( filterPass != 'PASS' and minAD != 0 and minRatioAD_DP != 0):
                        	#Filter if number of reads<10 if GT != 0/0
				if ((fdpnor < minAD and iaa != ['0', '0']) or (fdptum < minAD and iaar != ['0', '0'])):  
					#print 'Variant ' + str(ipos) + ' filtered by DP: ' + str(fdpnor) + '/' + str(fdptum) + ' in filename: ' + filename
					continue 
    			
				#Filter if the ratio of alternate reads < 0.05 if GT != 0/0
				if ((ratio_ad < minRatioAD_DP and iaa != ['0', '0']) or (ratio_adr < minRatioAD_DP and iaar != ['0', '0'])): 
					#print 'Variant ' + str(ipos) + ' filtered by AD: ' + str(ratio_ad) + '/' + str(ratio_adr) + ' in filename: ' + filename
					continue

			if gene.upper()=='UNKNOWN': continue

			# Include all type of deleterious mutations
			if filtMut:
				if v[ptype] != 'nonsynonymous SNV' and v[ptype] != 'nonframeshift insertion' and v[ptype] != 'nonframeshift deletion' and v[ptype] != 'frameshift insertion' and v[ptype] != 'frameshift deletion' and v[ptype] != 'stopgain' and v[ptype] != 'stoploss': continue 
			
			#print chr,ipos,aref,aalt,gene,'_'.join(v[ptype].split()),mut,freqMaxstr,''.join(iaa),ad_ref,ad_alt,''.join(iaar),adr_ref,adr_alt
			l.append([chr,ipos,aref,aalt,gene,'_'.join(v[ptype].split()),filterPass,freqMaxstr,''.join(iaa),ad_ref,ad_alt,''.join(iaar),adr_ref,adr_alt])	
	   try:
              fout=open(fileout, 'w')
           except: 
              print "Can't open ",fileout;
	   for row in l:
		fout.write('\t'.join([str(j) for j in row])+'\n')
	   fout.close()
	return 


def get_options():
	'''
    parse option from command line call
    '''
	import optparse

	desc = 'Script for filtering variants from annovar output'
	parser = optparse.OptionParser('usage: [-h] [-n normal] [-o outfile]', description=desc)
	parser.add_option('-n', dest='normpos', help='Position for checking germline mutation')
	parser.add_option('--no-germline', action='store_true', dest='rgerm', help='Filtering out germline mutations')
	parser.add_option('--both', action='store_true', dest='both', help='Filtering both positions')
	parser.add_option('-f', '--format', action='store', type='int', dest='pos_format',
					  help='Position of FORMAT information in VCF file')
	parser.add_option('-o', dest='outfile', help='Output file')
	parser.add_option('-b', '--bq', action='store', type='float', dest='bq', help='Base Quality Filter')
	parser.add_option('--fa', '--freq', action='store', type='float', dest='fa', help='Allele frequency Filter')
	parser.add_option('-d', '--depth', action='store', type='float', dest='dp', help='Depth Filter')
	parser.add_option('--no-pass', action='store_true', dest='npass', help='Remove PASS filtering')
	parser.add_option('--high', action='store_true', dest='hfvar', help='Select highly frequent variants')
	parser.add_option('--norm', action='store_true', dest='norm', help='Filter for normal positions')
	parser.add_option('--maf', '--maf_freq', action='store', type='float', dest='maf', help='MAF Minimum Allele frequency Filter')
	parser.add_option('--filtMut', action='store_true', dest='filtMut', help='Filter Synonimous variants')
	parser.add_option('--minAD', '--min_AD', action='store', type='float', dest='minAD', help='minAD Minimum Allele Depth')
	parser.add_option('--minRatioAD_DP', '--min_Ratio_AD_DP', action='store', type='float', dest='minRatioAD_DP', help='minRatioAD_DP Minimum Allele Ratio AD_DP')


	(options, args) = parser.parse_args()

	if (len(args) < 3):
		print 'Incorrect input'
		print 'python FilterAnnovarFiles.py path fileList  allele_column_number ' \
			  			'-n position_control'
		exit()
	else:
		if not os.path.isdir(args[0]):
			print 'ERROR:', args[0], 'input file not found'
			exit()

	rgerm = False
	both = False
	if options.normpos:
		try:
			normpos = int(options.normpos) - 1
			if options.rgerm: rgerm = True
			if options.both: both = True
		except:
			normpos = -1
	else:
		normpos = -1

	if options.norm:
		normFilt = True
	else:
		normFilt = False

	if options.npass:
		npass = False
	else:
		npass = True

	if options.bq:
		try:
			tbq = float(options.bq)
		except:
			tbq = 0
	else:
		tbq = 0

	if options.fa:
		try:
			tfa = float(options.fa)
		except:
			tfa = 0.0
	else:
		tfa = 0.0

	if options.dp:
		try:
			tdp = int(options.dp)
		except:
			tdp = 0
	else:
		tdp = 0

	if options.pos_format:
		global ann_tdat
		ann_tdat = int(options.pos_format) - 1

	if options.hfvar:
		low = False
	else:
		low = True

	if options.filtMut:
                filtMut = True
        else:
                filtMut = False

	if options.maf:
                try:
                        maf = float(options.maf)
                except:
                        maf = 1.0
        else:
                maf = 1.0

	if options.minAD:
                try:
                        minAD = float(options.minAD)
                except:
                        minAD = 0.0
        else:
                minAD = 0.0

	if options.minRatioAD_DP:
                try:
                        minRatioAD_DP = float(options.minRatioAD_DP)
                except:
                        minRatioAD_DP = 0.0
        else:
                minRatioAD_DP = 0.0

	opts = (normpos, both, rgerm, tdp, tfa, tbq, npass, low, normFilt, maf, filtMut, minAD, minRatioAD_DP)

	return args, opts


if __name__ == '__main__':
	global_vars()
	args, opts = get_options()
	if (len(sys.argv) > 4):
           path=sys.argv[1]
	   filelist=sys.argv[2]
	   icol = int(sys.argv[3]) - 1

	norm, both, rgerm, tdp, tfa, tbq, fpass, low, normFilt, maf, filtMut, minAD, minRatioAD_DP = opts
	if norm == icol:
		print >> sys.stderr, 'WARNING: Incorrect control position'
		norm = -1
	sel_vars = []

	get_annovar(path, filelist, icol, norm, tdp, tfa, tbq, fpass, both, normFilt, maf, filtMut, minAD, minRatioAD_DP)


